tests/unit_tests/test_coverage.py

```diff
@@ -34,21 +89,116 @@ def test_coverage_get_from_spades(metagenome, tmp_path):
-@pytest.mark.wip
-def test_coverage_get_from_bam(metagenome, tmp_path):
+def test_coverage_get_from_bam(metagenome, bam_alignment, tmp_path):
     out = tmp_path / "covs_from_bam.tsv"
-    df = coverage.get(fasta=metagenome, from_spades=False, out=out)
+    df = coverage.get(fasta=metagenome, from_spades=False, out=out, bam=bam_alignment)
     assert df.index.name == "contig"
     assert "coverage" in df.columns
     assert out.exists()
+
+
+def test_coverage_get_from_bed(metagenome, bed_alignment, tmp_path):
+    out = tmp_path / "covs_from_bed.tsv"
+    df = coverage.get(
+        fasta=metagenome, from_spades=False, out=out.as_posix(), bed=bed_alignment
+    )
+    assert df.index.name == "contig"
+    assert "coverage" in df.columns
+    assert out.exists()
+
+
+def test_coverage_get_from_reads(metagenome, fwd_reads, rev_reads, tmp_path):
+    out = tmp_path / "covs_from_bed.tsv"
+    df = coverage.get(
+        fasta=metagenome,
+        from_spades=False,
+        out=out.as_posix(),
+        fwd_reads=fwd_reads,
+        rev_reads=rev_reads,
+    )
+    assert df.index.name == "contig"
+    assert "coverage" in df.columns
+    assert out.exists()
+
+
+def test_get_ValueError(metagenome, tmp_path):
+    out = tmp_path / "covs.tsv"
+    with pytest.raises(ValueError):
+        coverage.get(fasta=metagenome, from_spades=False, out=out)
+
+
+def test_embed_df_already_exists(metagenome, df_exists_fpath, bed_alignment):
+    coverage.get(fasta=metagenome, out=df_exists_fpath, bed=bed_alignment)
+
+
+use_spades = [True, False]
+
+
+@pytest.fixture(name="mock_parser", params=use_spades)
+def fixture_mock_parser(
+    metagenome,
+    sam_alignment,
+    bam_alignment,
+    bed_alignment,
+    fwd_reads,
+    rev_reads,
+    monkeypatch,
+    tmp_path,
+    request,
+):
+    def return_mock_parser(*args, **kwargs):
+        return MockParser()
+
+    class MockParseArgs:
+        def __init__(
+            self,
+            metagenome,
+            sam_alignment,
+            bam_alignment,
+            bed_alignment,
+            fwd_reads,
+            rev_reads,
+            out,
+        ):
+            self.assembly = metagenome
+            self.fwd_reads = fwd_reads
+            self.rev_reads = rev_reads
+            self.sam = sam_alignment
+            self.bam = bam_alignment
+            self.lengths = "lengths.tsv"
+            self.bed = bed_alignment
+            self.cpus = 2
+            self.out = out
+            self.from_spades = request.param
+
+    class MockParser:
+        def add_argument(self, *args, **kwargs):
+            pass
+
+        def parse_args(self):
+            out = tmp_path / "binning.tsv"
+            return MockParseArgs(
+                metagenome,
+                sam_alignment,
+                bam_alignment,
+                bed_alignment,
+                fwd_reads,
+                rev_reads,
+                out,
+            )
+
+    # Defining the MockParser class to represent parser
```

Comment:
This comment should be above `class MockParser`

`monkeypatch.setattr(argparse, "ArgumentParser", return_mock_parser, raising=True)` is not defining the `MockParser` class, but patching `argparse.ArgumentParser()` with a function `return_mock_parser`.